opensearch-operator/pkg/reconcilers/securityconfig.go

```diff
@@ -0,0 +1,213 @@
+	for k := range data {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		_, err := hash.Write([]byte(key))
+		if err != nil {
+			return "", err
+		}
+		value := data[key]
+		_, err = hash.Write(value)
+		if err != nil {
+			return "", err
+		}
+	}
+	return base64.StdEncoding.EncodeToString(hash.Sum(nil)), nil
+}
+
+func (r *SecurityconfigReconciler) job(jobName string, namespace string, checksum string, adminCertName string) batchv1.Job {
+	dns := builders.DnsOfService(r.instance)
+	adminCert := "/certs/tls.crt"
+	adminKey := "/certs/tls.key"
+	caCert := "/certs/ca.crt"
+	var volumes []corev1.Volume
+	var volumeMounts []corev1.VolumeMount
+	volumes = append(volumes, corev1.Volume{
+		Name:         "securityconfig",
+		VolumeSource: corev1.VolumeSource{Secret: &corev1.SecretVolumeSource{SecretName: r.instance.Spec.Security.Config.SecurityconfigSecret.Name}},
+	})
+	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+		Name:      "securityconfig",
+		MountPath: "/securityconfig",
+	})
+	volumes = append(volumes, corev1.Volume{
+		Name:         "admin-cert",
+		VolumeSource: corev1.VolumeSource{Secret: &corev1.SecretVolumeSource{SecretName: adminCertName}},
+	})
+	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+		Name:      "admin-cert",
+		MountPath: "/certs",
+	})
+
+	arg := "ADMIN=/usr/share/opensearch/plugins/opensearch-security/tools/securityadmin.sh;" +
+		"chmod +x $ADMIN;" +
+		"count=0;" +
+		fmt.Sprintf("until $ADMIN -cacert %s -cert %s -key %s -cd /securityconfig/ -icl -nhnv -h %s.svc.cluster.local -p 9300 || (( count++ >= 20 )); do", caCert, adminCert, adminKey, dns) +
+		"  sleep 20; " +
+		"done"
+	annotations := map[string]string{
+		checksumAnnotation: checksum,
+	}
+	terminationGracePeriodSeconds := int64(5)
+	backoffLimit := int32(0)
+
+	return batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{Name: jobName, Namespace: namespace, Annotations: annotations},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: &backoffLimit,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{Name: jobName},
+				Spec: corev1.PodSpec{
+					TerminationGracePeriodSeconds: &terminationGracePeriodSeconds,
+					Containers: []corev1.Container{{
+						Name:         "updater",
+						Image:        builders.DockerImageForCluster(r.instance),
+						Command:      []string{"/bin/bash", "-c"},
+						Args:         []string{arg},
+						VolumeMounts: volumeMounts,
+					}},
+					Volumes:       volumes,
+					RestartPolicy: corev1.RestartPolicyNever,
+				},
+			},
+		},
+	}
+}
+
+func (r *SecurityconfigReconciler) determineAdminSecret() string {
+	if r.instance.Spec.Security.Config.AdminSecret.Name != "" {
+		return r.instance.Spec.Security.Config.AdminSecret.Name
+	} else if r.instance.Spec.Security.Tls.Transport.Generate {
+		return fmt.Sprintf("%s-admin-cert", r.instance.Name)
+	} else {
+		return ""
+	}
+}
+
+func (r *SecurityconfigReconciler) DeleteResources() (ctrl.Result, error) {
```

Comment:
also, maybe we should move it to helpers

Reply:
The `DeleteResources` method is intended to be part of every Reconciler and to be called by the controller when the custom resource is deleted. Its so the reconciler can delete/cleanup any objects it created that can't/shouldn't be deleted via a OwnerReference to the custom resource. At the moment it doesn't do anything as everything is cleaned up recursively via the OwnerReference.